README.md
# Ohjelmomistotekniikka, harjoitustyö

## Dokumentaatio

[Vaatimusmaarittely](documentaatio/vaatimusmaarittely.md)

[Arkkitehtuurikuvaus](documentaatio/arkkitehtuuri.md)

[Tuntikirjanpito](documentaatio/tuntikirjanpito.md)

[Changelog](documentaatio/changelog.md)

## Asennus

Asenna riippuvuudet komennolla:

    - poetry install


## Komentorivitoiminnot

### Käynistäminen

    poetry run invoke start

### Testaus

    poetry run invoke test

### Testikattavuus

    poetry run invoke coverage-report

### Pylint

    poetry run invoke lint

documentaatio/changelog.md
# Changelog

## Viikko3

-   Käyttäjä pystyy liikuttamaan lauta palasta oikealle 

-   Käyttäjä pystyy pyörittämään lauta palasta myötäpäivääm ja vastapäivään

-   Testattu, että lautapalasen liikkuminen toimii

## Viikko4

-   Lisättiin nuolet lauta palasen liikkuttamiseen

-   Pystyy sijoitamaan lauta palasen

-   Testattu, että luokka Polygon toimii kokonaisuudessa

src/tests/test_tile.py
import unittest
import numpy as np
from index import Hex
from index import RADIUS


class TestMaksukortti(unittest.TestCase):
    def setUp(self):
        self.hex = Hex(RADIUS, [0,0])

    def test_move(self):
        vector = [3, 4]
        self.hex.move(vector)

        points = np.array([
            [RADIUS*(np.sqrt(3)/2 + 3), RADIUS*(1/2 + 4)],
            [RADIUS*(0 + 3), RADIUS*(1 + 4)],
            [RADIUS*(-np.sqrt(3)/2 + 3), RADIUS*(1/2 + 4)],
            [RADIUS*(-np.sqrt(3)/2 + 3), RADIUS*(-1/2 + 4)],
            [RADIUS*(0 + 3), RADIUS*(-1 + 4)],
            [RADIUS*(np.sqrt(3)/2 + 3), RADIUS*(-1/2 + 4)]
        ])

        res = np.allclose(self.hex.points, points) and np.allclose(self.hex.dist_center, [RADIUS*3, RADIUS*4])

        self.assertEqual(res, True)


src/tests/test_polygon.py
import unittest
import numpy as np
import math

from logic import Polygon

POINTS = [
    [-10.0, 0.0],
    [10.0, 0.0],
    [10.0, 20.0],
    [-10.0, 20.0]
]


class TestPolygon(unittest.TestCase):
    def setUp(self):
        self.polygon = Polygon(POINTS)

    def test_edges(self):
        res_val = [
            [[-10.0, 0.0], [10.0, 0.0]],
            [[10.0, 0.0], [10.0, 20.0]],
            [[10.0, 20.0], [-10.0, 20.0]],
            [[-10.0, 20.0], [-10.0, 0.0]]
        ]

        res = True
        for edge, val in zip(self.polygon.edges, res_val):
            for (x, y), (x_res, y_res) in zip(edge, val):
                if not (math.isclose(x, x_res) and (y, y_res)):
                    res = False

        self.assertEqual(res, True)

    def test_move(self):
        vector = [10, 0]
        self.polygon.move(vector)

        res = np.allclose(self.polygon.points, [
            [0.0, 0.0],
            [20.0, 0.0],
            [20.0, 20.0],
            [0.0, 20.0]
        ])

        self.assertEqual(res, True)

    def test_rotate(self):
        self.polygon.rotate(90)

        res = np.allclose(self.polygon.points, [
            [0.0, -10.0],
            [0.0, 10.0],
            [-20.0, 10.0],
            [-20.0, -10.0]
        ])

        self.assertEqual(res, True)

    def test_point_not_in_polygon(self):
        point = [-1, -1]
        self.assertEqual(self.polygon.point_in_polygon(point), False)

    def test_point_in_polygon(self):
        point = [5, 5]
        self.assertEqual(self.polygon.point_in_polygon(point), True)

    def test_str(self):
        self.assertEqual(str(self.polygon), str(np.array(POINTS)))


laskarit/viikko2/unicafe/src/tests/kassapaate_test.py
import unittest
from kassapaate import kassapaate_data
from kassapaate import Kassapaate
from maksukortti import Maksukortti


class TestKassapaate(unittest.TestCase):
    def setUp(self):
        self.kassapaate = Kassapaate()
    
    def test_luotu_kassapaate(self):
        print(kassapaate_data(self.kassapaate))
        self.assertDictEqual(kassapaate_data(self.kassapaate), {
            'eurot': 1000,
            'sentit': 0,
            'edulliset': 0,
            'maukaat': 0
        })
    
    def test_maksuriittaa(self):
        edullinen_maksu = 300
        maukas_maksu = 500
        edullinen_vaihto = self.kassapaate.syo_edullisesti_kateisella(edullinen_maksu)
        maukas_vaihto = self.kassapaate.syo_maukkaasti_kateisella(maukas_maksu)

        kassa_data = kassapaate_data(self.kassapaate)
        kassa_data['edullinen_vaihto'] = edullinen_vaihto
        kassa_data['maukas_vaihto'] = maukas_vaihto

        self.assertDictEqual(kassa_data, {
            'eurot': 1006,
            'sentit': 40,
            'edulliset': 1,
            'maukaat': 1,
            'edullinen_vaihto': 60,
            'maukas_vaihto': 100
        })
    
    def test_maksu_ei_riita(self):
        edullinen_maksu = 200
        maukas_maksu = 300
        edullinen_vaihto = self.kassapaate.syo_edullisesti_kateisella(edullinen_maksu)
        maukas_vaihto = self.kassapaate.syo_maukkaasti_kateisella(maukas_maksu)

        kassa_data = kassapaate_data(self.kassapaate)
        kassa_data['edullinen_vaihto'] = edullinen_vaihto
        kassa_data['maukas_vaihto'] = maukas_vaihto

        self.assertDictEqual(kassa_data, {
            'eurot': 1000,
            'sentit': 0,
            'edulliset': 0,
            'maukaat': 0,
            'edullinen_vaihto': 200,
            'maukas_vaihto': 300
        })
    
    def test_maksu_riitta_kortti(self):
        kortti = Maksukortti(1000)

        edullinen_bool = self.kassapaate.syo_edullisesti_kortilla(kortti)
        maukas_bool = self.kassapaate.syo_maukkaasti_kortilla(kortti)

        kassa_data = kassapaate_data(self.kassapaate)
        kassa_data['kortin_saldo'] = kortti.saldo
        kassa_data['edullinen_bool'] = edullinen_bool
        kassa_data['maukas_bool'] = maukas_bool

        print(kassa_data)

        self.assertDictEqual(kassa_data, {
            'eurot': 1000,
            'sentit': 0,
            'edulliset': 1,
            'maukaat': 1,
            'kortin_saldo': 360,
            'edullinen_bool': True,
            'maukas_bool': True
        })

    def test_maksu_riitta_ei_kortti(self):
        kortti = Maksukortti(200)
        edullinen_bool = self.kassapaate.syo_edullisesti_kortilla(kortti)
        maukas_bool = self.kassapaate.syo_maukkaasti_kortilla(kortti)

        kassa_data = kassapaate_data(self.kassapaate)
        kassa_data['kortin_saldo'] = kortti.saldo
        kassa_data['edullinen_bool'] = edullinen_bool
        kassa_data['maukas_bool'] = maukas_bool

        self.assertDictEqual(kassa_data, {
            'eurot': 1000,
            'sentit': 0,
            'edulliset': 0,
            'maukaat': 0,
            'kortin_saldo': 200,
            'edullinen_bool': False,
            'maukas_bool': False
        })
    
    def test_lataa_rahaa(self):
        self.kortti = Maksukortti(0)
        self.kassapaate.lataa_rahaa_kortille(self.kortti, 300)
        self.assertListEqual([self.kassapaate.kassassa_rahaa, self.kortti.saldo], [100300, 300])
    
    def test_lataa_negatiivinen_rahaa(self):
        self.kortti = Maksukortti(0)
        self.kassapaate.lataa_rahaa_kortille(self.kortti, -1)
        self.assertListEqual([self.kassapaate.kassassa_rahaa, self.kortti.saldo], [100000, 0])


src/index.py
from __future__ import annotations
import sys
import pygame
import numpy as np

from logic import create_arrows
from logic import Tile

RED = 255, 0, 0
GREEN = 0, 255, 0
BLACK = 0, 0, 0
WHITE = 255, 255, 255

HEX_DIST_FROM_CENTER = 30
ARROW_SIZE = 70


WIDTH = 2*640
HEIGHT = 2*480

CENTER_POS = [WIDTH//2, HEIGHT//2]


def draw_polygon(screen, color, line_color, points):
    pygame.draw.polygon(screen, color, points + CENTER_POS)
    pygame.draw.polygon(screen, line_color, points + CENTER_POS, 2)


GRID = np.zeros((20, 20, 20))


def draw_grid(screen):
    points = []
    theta = np.radians(60)
    for i in range(6):
        x_pos = HEX_DIST_FROM_CENTER*np.cos(i*theta + np.radians(30))
        y_pos = HEX_DIST_FROM_CENTER*np.sin(i*theta + np.radians(30))
        points.append([x_pos, y_pos])

    points = np.array(points)

    for i in range(-10, 10):
        for j in range(-10, 10):
            for k in range(-10, 10):
                x_center = 2*(HEX_DIST_FROM_CENTER - 4) * \
                    (i + j*np.cos(theta) - k*np.cos(theta)) + WIDTH//2
                y_center = 2*(HEX_DIST_FROM_CENTER - 4) * \
                    (j*np.sin(theta) + k*np.sin(theta)) + HEIGHT//2

                pygame.draw.polygon(
                    screen, WHITE, points + [x_center, y_center])
                pygame.draw.polygon(screen, BLACK, points +
                                    [x_center, y_center], 2)


def main():
    pygame.init()

    screen = pygame.display.set_mode((WIDTH, HEIGHT))
    clock = pygame.time.Clock()

    draw_grid(screen)

    tile = Tile(HEX_DIST_FROM_CENTER)

    arrows = create_arrows(0, ARROW_SIZE)
    for arrow in arrows:
        arrow.move([-400, 280])

    while True:

        for hexagon in tile.hexes:
            draw_polygon(screen, WHITE, BLACK, hexagon.points)

        mouse_pos_x, mouse_pos_y = pygame.mouse.get_pos()

        for event in pygame.event.get():

            if event.type == pygame.QUIT:
                sys.exit()

            if event.type == pygame.KEYDOWN:

                if event.key == pygame.K_RIGHT:
                    tile.rotate(60)

                if event.key == pygame.K_LEFT:
                    tile.rotate(-60)

                if event.key == pygame.K_SPACE:
                    if tile.test_cordinates((GRID)):
                        tile.place(GRID)
                        for hexagon in tile.hexes:
                            draw_polygon(screen, GREEN, RED, hexagon.points)
                        tile = Tile(HEX_DIST_FROM_CENTER)

            if event.type == pygame.MOUSEBUTTONDOWN:
                for arrow in arrows:
                    arrow.click([mouse_pos_x - CENTER_POS[0],
                                mouse_pos_y - CENTER_POS[1]], tile)

        for hexagon in tile.hexes:
            draw_polygon(screen, GREEN, RED, hexagon.points)

        for arrow in arrows:
            if arrow.point_in_polygon([mouse_pos_x - CENTER_POS[0], mouse_pos_y - CENTER_POS[1]]):
                draw_polygon(screen, GREEN, RED, arrow.points)
            else:
                draw_polygon(screen, RED, GREEN, arrow.points)

        pygame.display.update()
        clock.tick(60)


if __name__ == "__main__":
    main()


src/logic/tile.py
import numpy as np
from .polygon import Polygon

HEX_MOVE_VECTORS = np.array([
    [np.cos(np.radians(30*i)), np.sin(np.radians(30*i))] for i in range(12)
])

MOVES = np.array([
    [1, 0, 0],
    [1, 0, 0],
    [0, 1, 0],
    [0, 1, 0],
    [0, 0, 1],
    [0, 0, 1],
    [-1, 0, 0],
    [-1, 0, 0],
    [0, -1, 0],
    [0, -1, 0],
    [0, 0, -1],
    [0, 0, -1]

])


class Hex(Polygon):
    def __init__(self, points_dist_center, center):

        points = []
        theta = np.radians(60)
        for i in range(6):
            x_pos = points_dist_center * \
                np.cos(i*theta + np.radians(30)) + points_dist_center*center[0]
            y_pos = points_dist_center * \
                np.sin(i*theta + np.radians(30)) + points_dist_center*center[1]
            points.append([x_pos, y_pos])

        self.points_dist_center = points_dist_center

        self.moves = np.array([0, 0, 0])

        self.valid_indexes = [0, 2, 4, 6, 8, 10]

        super().__init__(points)

    def move(self, vector):
        if vector in self.valid_indexes:
            self.moves += MOVES[vector]
            super().move(HEX_MOVE_VECTORS[vector]
                         * 2*(self.points_dist_center - 4))

    def check_cordinates(self, grid):
        if grid[self.moves[0]][self.moves[1]][self.moves[2]] == 0:
            return True
        return False

    def place(self, grid):
        grid[self.moves[0]][self.moves[1]][self.moves[2]] = 1


class Tile:
    def __init__(self, radius):
        vectors = np.array([
            [0.0, 0.0],
            [np.sqrt(3)/2, 3/2],
            [np.sqrt(3), 0.0],
            [np.sqrt(3)/2, -3/2]
        ])

        self.hexes = [
            Hex(radius, vector) for vector in vectors
        ]

    def move(self, index):
        for hexagon in self.hexes:
            hexagon.move(index)

    def rotate(self, theta):
        for hexagon in self.hexes:
            hexagon.rotate(theta)

    def test_cordinates(self, grid):
        return all(hexagon.check_cordinates(grid) for hexagon in self.hexes)

    def place(self, grid):
        for hexagon in self.hexes:
            hexagon.place(grid)

    def point_in_tile(self, point):
        if any(hexagon.point_in_polygon(point) for hexagon in self.hexes):
            return True
        return False


src/logic/arrow.py
import numpy as np


from .polygon import rotation_matrix
from .polygon import Polygon


BASIC_ARROW = 20 * np.array([
    [0, -1],
    [0, 1],
    [3, 1],
    [2, 3],
    [6, 0],
    [2, -3],
    [3, -1]
])


class Arrow(Polygon):
    def __init__(self, center, theta):
        self.center = center

        self.index = theta//30

        r_matrix = rotation_matrix(theta)
        points = []
        for point in BASIC_ARROW:
            rotated_point = r_matrix @ point + center
            points.append(rotated_point)

        super().__init__(points)

    def click(self, point, tile):
        if self.point_in_polygon(point):
            tile.move(self.index)


def create_arrows(theta, scalar):
    theta_rad = np.radians(theta)
    delta_theta = np.radians(60)
    return [
        Arrow([scalar*np.cos(delta_theta*i + theta_rad), scalar*np.sin(delta_theta*i+theta_rad)], +
              60*i + theta) for i in range(6)
    ]


src/logic/polygon.py
import numpy as np


def rotation_matrix(theta):
    theta_radians = np.radians(theta)
    return np.array([
        [np.cos(theta_radians), -np.sin(theta_radians)],
        [np.sin(theta_radians), np.cos(theta_radians)]
    ])


class Polygon:
    def __init__(self, points):
        self.points = np.array(points)

        self.center_vector = np.array([0.0, 0.0])

    @property
    def edges(self):
        res = []
        for point, next_point in zip(self.points, self.points[1:]):
            res.append([point, next_point])
        res.append([self.points[-1], self.points[0]])
        return res

    def move(self, vector):
        self.center_vector += vector
        self.points += vector

    def rotate(self, theta):
        r_matrix = rotation_matrix(theta)
        for i, point in enumerate(self.points):
            self.points[i] = r_matrix @ (point - self.center_vector) + \
                self.center_vector

    def point_in_polygon(self, point, epsilon=3):
        # epsilon is error term as pixels
        x_pos = point[0]
        y_pos = point[1]

        min_x = min(p[0] for p in self.points)
        max_x = max(p[0] for p in self.points)
        min_y = min(p[1] for p in self.points)
        max_y = max(p[1] for p in self.points)

        if (x_pos < min_x or x_pos > max_x) or (y_pos < min_y or y_pos > max_y):
            return False

        count = 0
        for (point1, point2) in self.edges:
            min_y, max_y = min(point1[1], point2[1]), max(point1[1], point2[1])
            if min_y < y_pos < max_y:
                delta_x = point1[0] - point2[0]
                delta_y = point1[1] - point2[1]
                if x_pos < (delta_x/delta_y)*(y_pos - point1[1]) + point1[0] + epsilon:
                    count += 1
        if count % 2 == 1:
            return True
        return False

    def __str__(self):
        return str(self.points)


laskarit/viikko2/unicafe/src/kassapaate.py
class Kassapaate:
    def __init__(self):
        self.kassassa_rahaa = 100000
        self.edulliset = 0
        self.maukkaat = 0

    def syo_edullisesti_kateisella(self, maksu):
        if maksu >= 240:
            self.kassassa_rahaa = self.kassassa_rahaa + 240
            self.edulliset += 1
            return maksu - 240
        else:
            return maksu

    def syo_maukkaasti_kateisella(self, maksu):
        if maksu >= 400:
            self.kassassa_rahaa = self.kassassa_rahaa + 400
            self.maukkaat += 1
            return maksu - 400
        else:
            return maksu

    def syo_edullisesti_kortilla(self, kortti):
        if kortti.saldo >= 240:
            kortti.ota_rahaa(240)
            self.edulliset += 1
            return True
        else:
            return False

    def syo_maukkaasti_kortilla(self, kortti):
        if kortti.saldo >= 400:
            kortti.ota_rahaa(400)
            self.maukkaat += 1
            return True
        else:
            return False

    def lataa_rahaa_kortille(self, kortti, summa):
        if summa >= 0:
            kortti.lataa_rahaa(summa)
            self.kassassa_rahaa += summa
        else:
            return
    
    def kassanraha_euroina(self):
        eurot = self.kassassa_rahaa//100
        sentit = self.kassassa_rahaa - eurot*100
        return {
            'eurot': eurot,
            'sentit': sentit
        }

def kassapaate_data(kassapaate:Kassapaate):
    kassanraha = kassapaate.kassanraha_euroina()
    eurot = kassanraha['eurot']
    sentit = kassanraha['sentit']
    eudulliset = kassapaate.edulliset
    maukaat = kassapaate.maukkaat
    return {
        'eurot':  eurot,
        'sentit': sentit,
        'edulliset': eudulliset,
        'maukaat': maukaat
    }